File: dictaite_core/services/text_utils.py
"""Utilities shared by transcription and translation services."""

from __future__ import annotations

import re

_PARA_SPLIT = re.compile(r"\n\s*\n")
_SPACE_COLLAPSE = re.compile(r"\s+")


def format_structured_text(text: str) -> str:
    """Normalise whitespace and paragraph spacing for readability."""

    stripped = text.strip()
    if not stripped:
        return ""
    paragraphs: list[str] = []
    for block in _PARA_SPLIT.split(stripped):
        block = block.strip()
        if not block:
            continue
        lines = [line.strip() for line in block.splitlines() if line.strip()]
        normalized = " ".join(_SPACE_COLLAPSE.sub(" ", line) for line in lines)
        if normalized:
            paragraphs.append(normalized)
    return "\n\n".join(paragraphs)


__all__ = ["format_structured_text"]

File: src/text_utils.rs
use once_cell::sync::Lazy;
use regex::Regex;

static PARA_SPLIT: Lazy<Regex> = Lazy::new(|| Regex::new(r"\n\s*\n").unwrap());
static SPACE_COLLAPSE: Lazy<Regex> = Lazy::new(|| Regex::new(r"\s+").unwrap());

pub fn format_structured_text(text: &str) -> String {
    let trimmed = text.trim();
    if trimmed.is_empty() {
        return String::new();
    }

    let mut paragraphs = Vec::new();
    for block in PARA_SPLIT.split(trimmed) {
        let block = block.trim();
        if block.is_empty() {
            continue;
        }
        let mut lines = Vec::new();
        for line in block.lines() {
            let line = line.trim();
            if line.is_empty() {
                continue;
            }
            let collapsed = SPACE_COLLAPSE.replace_all(line, " ");
            lines.push(collapsed);
        }
        if lines.is_empty() {
            continue;
        }
        paragraphs.push(lines.join(" "));
    }

    paragraphs.join("\n\n")
}
